diff --git a/mingw-w64-v7.0.0/mingw-w64-headers/crt/ctype.h b/mingw-w64-v7.0.0/mingw-w64-headers/crt/ctype.h
index 9d09833..9b45a9e 100644
--- a/mingw-w64-v7.0.0/mingw-w64-headers/crt/ctype.h
+++ b/mingw-w64-v7.0.0/mingw-w64-headers/crt/ctype.h
@@ -193,7 +193,7 @@ int __cdecl iswblank(wint_t _C);
 #ifndef _CTYPE_DISABLE_MACROS
 
 #ifndef MB_CUR_MAX
-#define MB_CUR_MAX ___mb_cur_max_func()
+#define MB_CUR_MAX __mb_cur_max
 #ifndef __mb_cur_max
 #ifdef _MSVCRT_
   extern int __mb_cur_max;
@@ -202,7 +202,7 @@ int __cdecl iswblank(wint_t _C);
 #ifndef _UCRT
   extern int * __MINGW_IMP_SYMBOL(__mb_cur_max);
 #endif
-#define __mb_cur_max	(___mb_cur_max_func())
+#define __mb_cur_max	__mb_cur_max
 #endif
 #endif
 _CRTIMP int __cdecl ___mb_cur_max_func(void);
diff --git a/mingw-w64-v7.0.0/mingw-w64-headers/crt/stdlib.h b/mingw-w64-v7.0.0/mingw-w64-headers/crt/stdlib.h
index 7d758f3..e85ef93 100644
--- a/mingw-w64-v7.0.0/mingw-w64-headers/crt/stdlib.h
+++ b/mingw-w64-v7.0.0/mingw-w64-headers/crt/stdlib.h
@@ -106,7 +106,7 @@ extern "C" {
 #define RAND_MAX 0x7fff
 
 #ifndef MB_CUR_MAX
-#define MB_CUR_MAX ___mb_cur_max_func()
+#define MB_CUR_MAX __mb_cur_max
 #ifndef __mb_cur_max
 #ifdef _MSVCRT_
   extern int __mb_cur_max;
@@ -115,7 +115,7 @@ extern "C" {
 #ifndef _UCRT
   extern int * __MINGW_IMP_SYMBOL(__mb_cur_max);
 #endif
-#define __mb_cur_max	(___mb_cur_max_func())
+#define __mb_cur_max	__mb_cur_max
 #endif
 #endif
 _CRTIMP int __cdecl ___mb_cur_max_func(void);
diff --git a/mingw-w64-v7.0.0/mingw-w64-headers/include/tdi.h b/mingw-w64-v7.0.0/mingw-w64-headers/include/tdi.h
index 3120070..6c5298b 100644
--- a/mingw-w64-v7.0.0/mingw-w64-headers/include/tdi.h
+++ b/mingw-w64-v7.0.0/mingw-w64-headers/include/tdi.h
@@ -24,8 +24,8 @@
 #define __TDI_H
 
 #include "ntddtdi.h"
-#include "tdistat.h"
-#include "netpnp.h"
+#include "ddk/tdistat.h"
+#include "ddk/netpnp.h"
 
 #ifdef __cplusplus
 extern "C" {
diff --git a/mingw-w64-v7.0.0/mingw-w64-libraries/winpthreads/src/thread.c b/mingw-w64-v7.0.0/mingw-w64-libraries/winpthreads/src/thread.c
index 1b12edb..d214376 100644
--- a/mingw-w64-v7.0.0/mingw-w64-libraries/winpthreads/src/thread.c
+++ b/mingw-w64-v7.0.0/mingw-w64-libraries/winpthreads/src/thread.c
@@ -30,6 +30,7 @@
 #include "thread.h"
 #include "misc.h"
 #include "winpthread_internal.h"
+#include <pseh/pseh.h>
 
 static _pthread_v *__pthread_self_lite (void);
 
@@ -54,20 +55,6 @@ static size_t idListCnt = 0;
 static size_t idListMax = 0;
 static pthread_t idListNextId = 0;
 
-#if !defined(_MSC_VER) || defined (USE_VEH_FOR_MSC_SETTHREADNAME)
-static void *SetThreadName_VEH_handle = NULL;
-
-static LONG __stdcall
-SetThreadName_VEH (PEXCEPTION_POINTERS ExceptionInfo)
-{
-  if (ExceptionInfo->ExceptionRecord != NULL &&
-      ExceptionInfo->ExceptionRecord->ExceptionCode == EXCEPTION_SET_THREAD_NAME)
-    return EXCEPTION_CONTINUE_EXECUTION;
-
-  return EXCEPTION_CONTINUE_SEARCH;
-}
-#endif
-
 typedef struct _THREADNAME_INFO
 {
   DWORD  dwType;	/* must be 0x1000 */
@@ -89,23 +76,14 @@ SetThreadName (DWORD dwThreadID, LPCSTR szThreadName)
 
    infosize = sizeof (info) / sizeof (ULONG_PTR);
 
-#if defined(_MSC_VER) && !defined (USE_VEH_FOR_MSC_SETTHREADNAME)
-   __try
+   _SEH_TRY
      {
        RaiseException (EXCEPTION_SET_THREAD_NAME, 0, infosize, (ULONG_PTR *)&info);
      }
-   __except (EXCEPTION_EXECUTE_HANDLER)
+   _SEH_HANDLE
      {
      }
-#else
-   /* Without a debugger we *must* have an exception handler,
-    * otherwise raising an exception will crash the process.
-    */
-   if ((!IsDebuggerPresent ()) && (SetThreadName_VEH_handle == NULL))
-     return;
-
-   RaiseException (EXCEPTION_SET_THREAD_NAME, 0, infosize, (ULONG_PTR *) &info);
-#endif
+   _SEH_END;
 }
 
 /* Search the list idList for an element with identifier ID.  If
@@ -419,21 +397,10 @@ __dyn_tls_pthread (HANDLE hDllHandle, DWORD dwReason, LPVOID lpreserved)
 
   if (dwReason == DLL_PROCESS_DETACH)
     {
-#if !defined(_MSC_VER) || defined (USE_VEH_FOR_MSC_SETTHREADNAME)
-      if (lpreserved == NULL && SetThreadName_VEH_handle != NULL)
-        {
-          RemoveVectoredExceptionHandler (SetThreadName_VEH_handle);
-          SetThreadName_VEH_handle = NULL;
-        }
-#endif
       free_pthread_mem ();
     }
   else if (dwReason == DLL_PROCESS_ATTACH)
     {
-#if !defined(_MSC_VER) || defined (USE_VEH_FOR_MSC_SETTHREADNAME)
-      SetThreadName_VEH_handle = AddVectoredExceptionHandler (1, &SetThreadName_VEH);
-      /* Can't do anything on error anyway, check for NULL later */
-#endif
     }
   else if (dwReason == DLL_THREAD_DETACH)
     {
